packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1470 @@
+            _vault[validator].kghDelegators[msg.sender].kghShares[tokenIds[i]] -= kghShareForTokenId;
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        _initUndelegateKghBatch(validator, tokenIds, kroInKghs, kroShares, kghShares, kghAssetsToWithdraw);
+
+        emit KghBatchUndelegationInitiated(validator, msg.sender, tokenIds, kroShares, kghShares);
+    }
+
+    /**
+     * @notice Finalize all pending KRO undelegation and returns the amount of assets that the vault would
+     *         exchange for the amount of shares provided.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The amount of assets that the vault would exchange for the amount of shares provided.
+     */
+    function finalizeUndelegate(address validator) public returns (uint128) {
+        if (_vault[validator].asset.totalPendingKroShares == 0) {
+            return 0;
+        }
+
+        KroDelegator storage delegator = _vault[validator].kroDelegators[msg.sender];
+        uint256[] memory requestTimes = delegator.undelegateRequestTimes;
+        uint128 assets;
+        uint128 shares;
+
+        for (uint256 i = 0; i < requestTimes.length; ) {
+            if (requestTimes[i] + WITHDRAWAL_DELAY <= block.timestamp) {
+                uint128 sharesToBurn = delegator.undelegateRequests[requestTimes[i]];
+
+                unchecked {
+                    assets += sharesToBurn.mulDiv(
+                        _vault[validator].asset.totalPendingAssets,
+                        _vault[validator].asset.totalPendingKroShares
+                    );
+                    shares += sharesToBurn;
+                }
+
+                delete delegator.undelegateRequests[requestTimes[i]];
+                delete delegator.undelegateRequestTimes[i];
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        _vault[validator].asset.totalPendingAssets -= assets;
+        _vault[validator].asset.totalPendingKroShares -= shares;
+
+        bool success = IERC20(GOVERNANCE_TOKEN).transfer(msg.sender, assets);
+        require(success, "AssetManager: Transfer failed");
+
+        emit KroUndelegationFinalized(validator, msg.sender, assets);
+        return assets;
+    }
+
+    /**
+     * @notice Finalize all pending KGH undelegation and returns the amount of assets that the vault would
+     *         exchange for the amount of shares provided.
+     *
+     * @param validator Address of the validator.
+     *
+     * @return The amount of assets that the vault would exchange for the amount of shares provided.
+     */
+    function finalizeUndelegateKgh(address validator) external returns (uint128) {
+        KghDelegator storage kghDelegator = _vault[validator].kghDelegators[msg.sender];
+        uint256[] memory requestTimes = kghDelegator.undelegateRequestTimes;
+        uint128 assets = finalizeUndelegate(validator);
```

Comment:
Also if we remain as is, this line should be like below because all assets related to base reward is already handled in `finalizeUndelegate`.
```suggestion
        finalizeUndelegate(validator);
        uint128 assets;
```

Reply:
The first one looks better in terms of gas, so I'll consider applying it. Thanks!